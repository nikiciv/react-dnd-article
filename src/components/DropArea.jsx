/* @flow */
import React from 'react';

// * onFileDrop: a callback function that will receive an array of files as first argument
// * style: offers possibility to add styles
// * className: offers possibility to style the component according to the context
type Props = {
  onFileDrop: ({}) => void,
  style: object,
  className: string,
};

// This component will handle the file drop
class DropArea extends React.PureComponent<Props> {
  componentDidMount = () => {
    window.addEventListener('dragover', this.preventDefaultDragOver, false);
    window.addEventListener('drop', this.preventDefaultDrop, false);
  };

  componentWillUnmount = () => {
    window.removeEventListener('dragover', this.preventDefaultDragOver, false);
    window.removeEventListener('drop', this.preventDefaultDrop, false);
  };

  preventDefaultDrop = e => {
    e.preventDefault();
  };

  preventDefaultDragOver = e => {
    e.preventDefault();
  };

  handleDragOver = e => {
    const { dataTransfer } = e;
    e.stopPropagation();
    // Prevent default behavior(Prevent file from being opened)
    e.preventDefault();
    dataTransfer.dropEffect = 'copy';
  };

  handleFileDrop = e => {
    e.stopPropagation();
    e.preventDefault();
    if (this.props.onFileDrop) {
      this.props.onFileDrop(e.dataTransfer.files[0]);
    }
  };

  render() {
    return (
      <div
        style={this.props.style}
        onDragOver={this.handleDragOver}
        onDrop={this.handleFileDrop}
        className={this.props.className}
      >
        {this.props.children}
      </div>
    );
  }
}

export default DropArea;
