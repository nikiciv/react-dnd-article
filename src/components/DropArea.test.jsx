import React from 'react';
import { mount } from 'enzyme';
import DropArea from './DropArea';

describe('<DropArea />', () => {
  const props = {
    className: 'abc',
    onFileDrop: jest.fn(),
  };

  let instance;

  beforeEach(() => {
    instance = mount(<DropArea {...props} />);
  });

  it('renders DropArea component correctly', () => {
    expect(instance).toMatchSnapshot();
  });

  it('calls onFileDrop function when drop event is triggered', () => {
    const event = { dataTransfer: { files: ['test'] } };
    instance.simulate('drop', event);
    expect(props.onFileDrop).toBeCalledWith(event.dataTransfer.files[0]);
  });
});
