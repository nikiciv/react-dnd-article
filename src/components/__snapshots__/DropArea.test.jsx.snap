// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropArea /> renders DropArea component correctly 1`] = `
<DropArea
  className="abc"
  onFileDrop={[MockFunction]}
>
  <div
    className="abc"
    onDragOver={[Function]}
    onDrop={[Function]}
  />
</DropArea>
`;
